### ISSUE TITLE
`omp_is_initial_device` is non-constexpr at least on icpx

### ISSUE THREAD

**Issue by tom91136** [kokkos/kokkos#3661]
This is sorta related to #3335

While testing the experimental GEN9+OPENMPTARGET support, Intel's `icpx` in their oneAPI distribution doesn't seems to be a compile time constant. Skimming the OpenMP 5.0 spec, it's not obvious whether `omp_is_initial_device` has any special constraints other than it lives in `omp.h`.
We are building Kokkos in-tree using the method detailed in the build doc in case this changes things.

As an attempt to workaround this, I've removed the `constexpr` keyword in the condition (which drops the C++17 requirement) ) and re-added the `#include <omp.h>` line:

```cpp
// undefined at end of file
#if defined(KOKKOS_ENABLE_OPENMPTARGET)
#if defined(KOKKOS_COMPILER_PGI)
#define KOKKOS_IMPL_IF_ON_HOST if (!__builtin_is_device_code())
#else
#include <omp.h>
#define KOKKOS_IMPL_IF_ON_HOST if (omp_is_initial_device())
#endif
#else
#define KOKKOS_IMPL_IF_ON_HOST if (true)
#endif
```

After this our project compiles and seems to run correctly on the GPU.

Also, while we're testing this, we've also found that the script wasn't able to detect `icpx` properly (it keeps thinking it's `Clang` and not `IntelClang`), I've opened #3660 for this.

**Comment by crtrott** [comment]
This is weird, we definitely were testing with the Intel compiler too I ping some folks.

**Comment by rgayatri23** [comment]
We are trying to fix the issue with PR #3611 by doing this
```
#pragma omp begin declare variant match(device = {kind(host)})
static constexpr bool kokkos_omp_on_host() { return true; }
#pragma omp end declare variant

#pragma omp begin declare variant match(device = {kind(nohost)})
static constexpr bool kokkos_omp_on_host() { return false; }
#pragma omp end declare variant

#define KOKKOS_IMPL_IF_ON_HOST if constexpr (kokkos_omp_on_host())
```

**Comment by crtrott** [comment]
Apparently this is a change in the compiler.

**Comment by pvelesko** [comment]
@crtrott  @rgayatri23  I did a git pull on the Develop branch today and ran into this issue again.
Regression?

**Comment by crtrott** [comment]
Yeah, pr to fix it is already there.

**Comment by crtrott** [comment]
Pull again just merged the PR #4160
